File: skills/current/market-hunter/lib/adapters/kinguin.ts
import { Effect } from "effect";
import type { AdapterError, MarketplaceAdapter, RawMarketListing, SearchTarget } from "#models";
import { normalizeRawItems, type RawScrapedItem } from "./common.ts";

export class KinguinAdapter implements MarketplaceAdapter {
  readonly id = "kinguin" as const;
  readonly displayName = "Kinguin";
  readonly isEnabledByDefault = true;

  buildSearchTarget(query: string): SearchTarget {
    const encoded = encodeURIComponent(query.trim());
    return {
      marketplace: this.id,
      url: `https://www.kinguin.net/listing?phrase=${encoded}&active=1&hide_out_of_stock=1`,
      waitForMs: 2000,
      format: "json",
    };
  }

  parseListings(raw: unknown): Effect.Effect<readonly RawMarketListing[], AdapterError> {
    if (!raw) return Effect.succeed([]);

    if (typeof raw === "string") {
      const items: RawScrapedItem[] = [];
      const linkRegex = /\[([^\]]+)\]\((https:\/\/www\.kinguin\.net\/category\/[^)]+)\)[\s\S]*?([0-9]+(?:\.[0-9]+)?)\s*(?:USD|\$|EUR|€)/g;
      let match: RegExpExecArray | null;

      while ((match = linkRegex.exec(raw)) !== null) {
        const title = match[1]?.trim();
        const url = match[2];
        const price = match[3];
        if (title && price && !title.includes("logo") && !title.includes("Sign in")) {
          items.push({
            id: url ? url.split("/category/")[1]?.split("/")[0] || `kinguin-${items.length + 1}` : undefined,
            title,
            url,
            priceUsd: price,
            sellerRating: 97,
          });
        }
      }

      return normalizeRawItems(items, this.id, 30);
    }

    if (typeof raw === "object") {
      const obj = raw as Record<string, unknown>;
      const rawItems = Array.isArray(obj["items"])
        ? (obj["items"] as readonly RawScrapedItem[])
        : Array.isArray(obj["products"])
          ? (obj["products"] as readonly RawScrapedItem[])
          : [];

      return normalizeRawItems(rawItems, this.id, 30);
    }

    return Effect.succeed([]);
  }
}

File: skills/current/market-hunter/lib/adapters/kinguin.py
"""Kinguin marketplace adapter (port of lib/adapters/kinguin.ts)."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from adapters.common import RawScrapedItem, normalize_raw_items

if TYPE_CHECKING:
    from models import MarketplaceId, RawMarketListing, SearchTarget

_LINK_RE = re.compile(
    r"\[([^\]]+)\]\((https://www\.kinguin\.net/category/[^)]+)\)"
    r"[\s\S]*?([0-9]+(?:\.[0-9]+)?)\s*(?:USD|\$|EUR|€)"
)


class KinguinAdapter:
    """Kinguin marketplace adapter."""

    id: MarketplaceId = "kinguin"
    display_name = "Kinguin"
    is_enabled_by_default = True

    def build_search_target(self, query: str) -> SearchTarget:
        """Build the Kinguin listing URL for a query."""
        encoded = quote(query.strip(), safe="!'()*")
        return {
            "marketplace": self.id,
            "url": f"https://www.kinguin.net/listing?phrase={encoded}&active=1&hide_out_of_stock=1",
            "waitForMs": 2000,
            "format": "json",
        }

    def parse_listings(self, raw: Any) -> list[RawMarketListing]:
        """Parse Kinguin markdown or JSON scrape output into listings."""
        if not raw:
            return []

        if isinstance(raw, str):
            items: list[RawScrapedItem] = []
            for match in _LINK_RE.finditer(raw):
                title = match.group(1).strip() if match.group(1) else ""
                url = match.group(2)
                price = match.group(3)
                if title and price and "logo" not in title and "Sign in" not in title:
                    item: RawScrapedItem = {
                        "title": title,
                        "url": url,
                        "priceUsd": price,
                        "sellerRating": 97,
                    }
                    if url:
                        parts = url.split("/category/")
                        item_id = parts[1].split("/")[0] if len(parts) > 1 else ""
                        item["id"] = item_id or f"kinguin-{len(items) + 1}"
                    items.append(item)

            return normalize_raw_items(items, self.id, 30)

        if isinstance(raw, dict):
            raw_items = raw.get("items")
            if not isinstance(raw_items, list):
                raw_items = raw.get("products")
            return normalize_raw_items(
                raw_items if isinstance(raw_items, list) else [],
                self.id,
                30,
            )

        return []
